mixing/mixpool/observer.go

```diff
@@ -0,0 +1,503 @@
+// Copyright (c) 2025 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package mixpool
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/decred/dcrd/chaincfg/chainhash"
+	"github.com/decred/dcrd/wire"
+)
+
+const strikeLimit = 2
+
+const minPeers = 4 // Keep synced with mixclient.MinPeers
+
+type strikeSet struct {
+	set     map[wire.OutPoint]struct{}
+	strikes []uint64
+}
+
+func sortUniq(strikes []uint64) []uint64 {
+	sort.Slice(strikes, func(i, j int) bool {
+		return strikes[i] < strikes[j]
+	})
+	n := len(strikes)
+	strikes = strikes[:1]
+	last := strikes[0]
+	for _, epoch := range strikes[1:n] {
+		if epoch != last {
+			strikes = append(strikes, epoch)
+			last = epoch
+		}
+	}
+	return strikes
+}
+
+func (s *strikeSet) merge(other *strikeSet) {
+	if s == other {
+		return
+	}
+
+	for op := range other.set {
+		s.set[op] = struct{}{}
+	}
+
+	s.strikes = sortUniq(append(s.strikes, other.strikes...))
```

Comment:
And there is no guarantee that they're both non-empty thus calling `sortUniq` with an empty slice which can panic.